src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/CodeFixesForProjects.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using OmniSharp.Services;
+using OmniSharp.Utilities;
+
+namespace OmniSharp.Roslyn.CSharp.Services.Refactoring.V2
+{
+    [Shared]
+    [Export(typeof(CodeFixesForProjects))]
+    public class CodeFixesForProjects
+    {
+        private readonly ConcurrentDictionary<string, IEnumerable<CodeFixProvider>> _codeFixCache = new ConcurrentDictionary<string, IEnumerable<CodeFixProvider>>();
+        private readonly IAssemblyLoader _assemblyLoader;
+
+        [ImportingConstructor]
+        public CodeFixesForProjects(IAssemblyLoader assemblyLoader)
+        {
+            _assemblyLoader = assemblyLoader;
+        }
+
+        public IEnumerable<CodeFixProvider> GetAllCodeFixesForProject(string projectId)
+        {
+            if (_codeFixCache.ContainsKey(projectId))
+                return _codeFixCache[projectId];
+            return Enumerable.Empty<CodeFixProvider>();
+        }
+
+        public void LoadFrom(string projectId, IEnumerable<string> AnalyzerPaths)
+        {
+            var codeFixes = AnalyzerPaths
+                .Where(x => x.EndsWith("CodeFixes.dll"))
+                .SelectMany(codeFixDllPath =>
+                {
+                    var loadedAssembly = _assemblyLoader.LoadFrom(codeFixDllPath);
+                    var validTypes = loadedAssembly.GetTypes()
+                        .Where(type => !type.GetTypeInfo().IsInterface && !type.GetTypeInfo().IsAbstract && !type.GetTypeInfo().ContainsGenericParameters)
+                        .Where(t => typeof(CodeFixProvider).IsAssignableFrom(t));
+
+                    return validTypes
+                        .Select(type => type.CreateInstance<CodeFixProvider>())
+                        .Where(instance => instance != null);
+                });
+
```

Comment:
same question as RuleSet. what is lifetime management? will vscode create new omnisharp server when new solution is opened or something? is that the reason there is no remove?

Reply:
what happen if user update those dlls from disk while things are opened in VS? like nuget update or etc? the existing assembly loader lock the file so, it won't work anyway.. but is that expected behavior?

Reply:
cc @david-driscoll 

I am not sure about lifetime. I think omnisharp supports only one solution and is re-created between solutions?